cores/esp8266/StreamDev.cpp

```diff
@@ -0,0 +1,333 @@
+
+    This library is free software; you can redistribute it and/or
+    modify it under the terms of the GNU Lesser General Public
+    License as published by the Free Software Foundation; either
+    version 2.1 of the License, or (at your option) any later version.
+
+    This library is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+    Lesser General Public License for more details.
+
+    You should have received a copy of the GNU Lesser General Public
+    License along with this library; if not, write to the Free Software
+    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+
+    parsing functions based on TextFinder library by Michael Margolis
+*/
+
+
+#include <Arduino.h>
+#include <StreamDev.h>
+
+using esp8266::polledTimeout::oneShotFastMs;
+using esp8266::polledTimeout::periodicFastMs;
+
+size_t Stream::toFull(Print* to,
+                      const ssize_t len,
+                      const int readUntilChar,
+                      const oneShotFastMs::timeType timeoutMs)
+{
+    setWriteError(STREAMTO_SUCCESS);
+
+    if (len == 0)
+    {
+        return 0;    // conveniently avoids timeout for no requested data
+    }
+
+    // There are two timeouts:
+    // - read (network, serial, ...)
+    // - write (network, serial, ...)
+    // However
+    // - getTimeout() is for reading only
+    // - there is no getOutputTimeout() api
+    // So we use getTimeout() for both,
+    // (also when inputTimeoutPossible() is false)
+
+    // "neverExpires (default, impossible)" is translated to default timeout
+    oneShotFastMs timedOut(timeoutMs >= oneShotFastMs::neverExpires ? getTimeout() : timeoutMs);
+
+    // yield about every 5ms (XXX SHOULD BE A SYSTEM-WIDE CONSTANT?)
+    periodicFastMs yieldNow(5);
+
+    size_t written = 0;
+
+    // len==-1 => maxLen=0 <=> until starvation
+    size_t maxLen = std::max((ssize_t)0, len);
+
+    if (peekBufferAPI())
+
+        // peek-buffer API
+
+        while (!maxLen || written < maxLen)
+        {
+            size_t avpk = availableForPeek();
+            if (avpk == 0 && !inputTimeoutPossible())
+            {
+                // no more data to read, ever
+                break;
+            }
+
+            size_t w = to->availableForWrite();
+            if (w == 0 && !outputTimeoutPossible())
+            {
+                // no more data can be written, ever
+                break;
+            }
+
+            w = std::min(w, avpk);
+            if (maxLen)
+            {
+                w = std::min(w, maxLen - written);
+            }
+            if (w)
+            {
+                const char* directbuf = peekBuffer();
+                bool foundChar = false;
+                if (readUntilChar >= 0)
+                {
+                    const char* last = (const char*)memchr(directbuf, readUntilChar, w);
```

Comment:
Do we need memchr_P?  Same as other question, does the peekBuffer API enforce byte-addressible RAM only?

Reply:
Currently the peekBuffer API enforces byte-addressible RAM only, like Stream API.
`directbuf` is then used in `to->write()` which is `Stream::write()`.